src/Auth/AuthProvider.php

```diff
@@ -0,0 +1,333 @@
+            );
+        }
+        if (empty($json['access_token']) || empty($json['expires_in'])) {
+            throw new AuthorizationFailedException(
+                sprintf('An error occurred while %s flow. Message: invalid response', $authType->value)
+            );
+        }
+    }
+
+    /**
+     * @throws AuthorizationFailedException
+     */
+    public function oauth2Refresh(): ?Credentials
+    {
+        if ($this->credentials !== null) {
+            try {
+                $json = $this->makeRequest([
+                    'grant_type' => 'refresh_token',
+                    'client_id' => $this->clientId,
+                    'client_secret' => $this->clientSecret,
+                    'refresh_token' => $this->credentials->getRefreshToken(),
+                    'redirect_uri' => $this->redirectUri,
+                ]);
+            } catch (InvalidDataException | ConnectionFailedException $exception) {
+                throw new AuthorizationFailedException(
+                    sprintf('Error occurred while refreshing token. Message: %s', $exception->getMessage()),
+                    $exception->getCode(),
+                    $exception
+                );
+            } catch (RequestException $exception) {
+                throw new AuthorizationFailedException(
+                    sprintf('Error occurred. Message: %s', $exception->getResponse()->getReasonPhrase()),
+                    $exception->getCode(),
+                    $exception
+                );
+            }
+
+            $authType = AuthTypeEnum::AUTHORIZATION_CODE_FLOW;
+            $this->checkResponseWithAccessToken($json, $authType);
+            /** @var array{'refresh_token'?: string|null, 'access_token': string, 'expires_in': int} $json */
+            $this->credentials = new Credentials(
+                $json['refresh_token'] ?? null,
+                $json['access_token'],
+                (new \DateTimeImmutable())->modify('+ ' . ($json['expires_in'] - 10) . ' seconds'),
+                $authType
+            );
+            $this->callCredentialsCallback();
+            return $this->credentials;
+        }
+        return $this->credentials;
+    }
+
+    /**
+     * @throws AuthorizationFailedException
+     * @throws ClientErrorException
+     * @throws ClientExceptionInterface
+     * @throws ServerErrorException
+     */
+    public function revoke(): bool
+    {
+        if ($this->credentials === null) {
+            throw new AuthorizationFailedException('Load authentication screen first.');
+        }
+        if ($this->credentials->getAuthType()->value !== AuthTypeEnum::AUTHORIZATION_CODE_FLOW->value) {
+            throw new AuthorizationFailedException('Revoke is only available for authorization code flow');
+        }
+        try {
+            $request = new Request(
+                'POST',
+                sprintf('%s/oauth/revoke?token=' . $this->credentials->getRefreshToken(), Dispatcher::BASE_URL),
```

Comment:
Token by měl jít také v json body.

Reply:
vím, měl jsem to upravené, ale ještě jsem to nepushnul :) Už to tu je